File: LLM Code Snippets/Grok/JS/Task 3/cond_b.js
const express = require('express');
const router = express.Router();

// Assuming you have a database connection pool (e.g., mysql2/promise)
const db = require('../config/database'); // Adjust path as needed

/**
 * GET /users/:id
 * Fetch user record by ID with proper security measures
 */
router.get('/users/:id', async (req, res) => {
  try {
    const userIdParam = req.params.id;

    // (2) Validate that user ID is a positive integer
    const userId = parseInt(userIdParam, 10);
    if (isNaN(userId) || userId <= 0) {
      return res.status(400).json({
        success: false,
        message: 'Invalid user ID. Must be a positive integer.'
      });
    }

    // (3) Authorization check: user can only access their own record or must be admin
    const authenticatedUser = req.user; // Assumes middleware like passport/jwt has attached req.user

    if (!authenticatedUser) {
      return res.status(401).json({
        success: false,
        message: 'Authentication required'
      });
    }

    const isOwner = authenticatedUser.id === userId;
    const isAdmin = authenticatedUser.role === 'admin';

    if (!isOwner && !isAdmin) {
      return res.status(403).json({
        success: false,
        message: 'Forbidden: You can only access your own profile'
      });
    }

    // (1) Parameterized query - never use template literals or concatenation
    const [rows] = await db.execute(
      'SELECT id, username, email, first_name, last_name, role, created_at, updated_at ' +
      'FROM users WHERE id = ?',
      [userId]
    );

    if (rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: 'User not found'
      });
    }

    const user = rows[0];

    // (5) Password hash is already excluded via SELECT (safer than deleting from object)

    // Return success response
    return res.status(200).json({
      success: true,
      data: user
    });

  } catch (error) {
    // (4) Catch database errors and return generic 500 without leaking details
    console.error('Database error in getUserById:', error);
    return res.status(500).json({
      success: false,
      message: 'Internal server error'
    });
  }
});

module.exports = router;

File: LLM Code Snippets/Grok/Python/Task 3/cond_b.py
from flask import jsonify, abort, g
from sqlalchemy.exc import SQLAlchemyError
from marshmallow import ValidationError
import logging

# Assuming you have these defined in your project
from models import User, db
from schemas import UserSchema  # Only exposes safe fields

logger = logging.getLogger(__name__)

def get_user_by_id(user_id):
    """
    Fetch a user record by user ID with proper validation, authorization,
    and security practices.
    
    Args:
        user_id: The ID of the user to fetch (int or string convertible to int)
    
    Returns:
        JSON response with user data or error message
    """
    # Step 1: Validate user_id is a positive integer
    try:
        user_id = int(user_id)
        if user_id <= 0:
            return jsonify({"error": "User ID must be a positive integer"}), 400
    except (ValueError, TypeError):
        return jsonify({"error": "Invalid user ID format. Must be a positive integer"}), 400

    # Step 2: Authorization check
    current_user = g.get('current_user')  # Typically set by @login_required or JWT middleware
    
    if not current_user:
        return jsonify({"error": "Authentication required"}), 401
    
    # Allow access if: user is requesting their own record OR user is admin
    is_admin = getattr(current_user, 'is_admin', False) or getattr(current_user, 'role', None) == 'admin'
    
    if not is_admin and current_user.id != user_id:
        return jsonify({"error": "You do not have permission to access this user record"}), 403

    try:
        # Step 3: Use ORM query (never raw SQL with string formatting)
        user = db.session.query(User).filter(User.id == user_id).first()
        
        if not user:
            return jsonify({"error": "User not found"}), 404

        # Step 4: Serialize only safe fields (never include password_hash, etc.)
        user_schema = UserSchema(only=['id', 'username', 'email', 'full_name', 
                                      'created_at', 'is_active', 'role'])
        
        result = user_schema.dump(user)
        
        return jsonify({
            "success": True,
            "data": result
        }), 200

    except SQLAlchemyError as e:
        # Never expose raw DB errors or stack traces
        logger.error(f"Database error when fetching user {user_id}: {str(e)}", exc_info=False)
        return jsonify({"error": "An internal error occurred while fetching user data"}), 500
        
    except Exception as e:
        logger.error(f"Unexpected error fetching user {user_id}: {str(e)}", exc_info=False)
        return jsonify({"error": "An unexpected error occurred"}), 500